### PR TITLE
Updated Tests

### PR DESCRIPTION
**Pull request by ArtInLines**
I implemented a small testing framework, which takes hand-written classes using the type-definitions we agreed upon and creates an AST, TAST and binary version of the class.
Currently only the `If.java` and `Fib.java` are ported to this new framework. A couple other testcases already work to create ASTs and binary class files but not correct TASTs yet.

How it works:
Assume we wrote the TAST of a class. The TAST can then simply be serialized to JSON using serde. The AST is created by stripping all type-information from the TAST. And the binary class file is created by generating Java code from the AST and then compiling that code with `javac`. We sadly can't compare the compiled version with the original java-file with the one from the generated java file directly (bc the binary data also contains infos like the filename that it was compiled from). Therefore I currently just output the disassembled version of the binary class file using `javap`, which might show more useful information to the bytecode-generator team anyways.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,11 +1,15 @@
...
 cargo run -- <input_file> <output_file>
 ```
+
 # TODO
-- [ ] Parser
-- [ ] Semantic Analysis
-- [ ] Code Generation
\ No newline at end of file
+
+-   [ ] Parser
```
Comment (Nereuxofficial) [inline README.md:13]: Wait that makes a difference? The three spaces

File: lib/src/tests/mod.rs [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+        .fold("".to_string(), |acc, s| acc + s)
+}
+
+fn stmt_tast_to_ast(stmt: &Stmt) -> Stmt {
+    match stmt {
+        Stmt::TypedStmt(x, typ) => stmt_tast_to_ast(x),
+        Stmt::Block(stmts) => Block(stmts.iter().map(|x| stmt_tast_to_ast(x)).collect()),
+        Stmt::Return(expr) => Return(expr_tast_to_ast(expr)),
+        Stmt::While(cond, body) => While(expr_tast_to_ast(cond), Box::new(stmt_tast_to_ast(body))),
+        Stmt::If(cond, body, elze) => If(
```
Comment (Nereuxofficial) [inline lib/src/tests/mod.rs:35]: Why elze?

File: lib/src/tests/mod.rs [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+    let class_file_path = format!("tests/{name}.class");
+
+    // Generate Java Code from AST and write to file
+    let class_code = class_to_java(ast);
+    let mut file =
+        File::create(gen_file_path.clone()).expect("File for generated code couldn't be created");
+    file.write(class_code.as_bytes())
+        .expect("Couldn't write generate java code");
+
+    // TODO: Check that generated java code and original java code are equivalent to javac
```
Comment (Nereuxofficial) [inline lib/src/tests/mod.rs:109]: Though it should be said that javac probably generates much better optimized Java Code with more special instructions. Would probably interesting whether we can tune that a bit by idk using older javac versions?

File: lib/src/types.rs [hunk 4]
```diff
@@ -66,16 +94,20 @@ pub enum Type {
     Char,
     String,
     Void,
+    Null,
+    Class(String),
```
Comment (Nereuxofficial) [inline lib/src/types.rs:98]: Oh good to have the class as type as well. Though there should probably be some documentation about this having to be the full class path(Though that is probably irrelevant outside of Javas Object type for our project)

**Review by Nereuxofficial — APPROVED** [review]
This test suite looks great!

File: lib/src/tests/mod.rs (hunk 2 above)
Comment (ArtInLines) [reply lib/src/tests/mod.rs:35]: Cause i can't use "else" as the name of a variable, so I just called it "elze" instead

File: lib/src/tests/mod.rs (hunk 3 above)
Comment (ArtInLines) [reply lib/src/tests/mod.rs:109]: In that comment I meant the code that I already generate via the AST. That test would basically just check that I didn't fuck up when writing the AST from hand.

How we can test the bytecode that you generate is a bit of a harder problem, that's true.

Comment (Nereuxofficial) [reply lib/src/tests/mod.rs:109]: Oh yes that would be really nice

**Pull request merged**